Pretty printer drops variance from LHS of type substitution

<!--
Welcome to OCaml's Issue tracker!

OCaml's developers use this tracker for bugs and feature requests only, rather
than user support.

If you have questions about *using* OCaml, please ask at
https://discuss.ocaml.org (more people read Discuss than this tracker, and
you'll get confirmation of whether you've really found a bug or need a new
feature).

If your error came from the OCaml package manager, opam, (messages beginning
`[ERROR] The compilation of ...`), please start at
https://github.com/ocaml/opam-repository/issues/new.

Some libraries and tools which used to be part of OCaml are now maintained
separately. Please post questions about Graphics, Num, camlp4, LablTk, CamlDBM
or OCamlTopWin on Discuss or on their respective issue trackers (see [README.adoc](https://github.com/ocaml/ocaml/blob/trunk/README.adoc#separately-maintained-components)
for a full list).
-->

I was surprised to discover this syntax is valid, but apparently so:

```
module type T1 = sig
  type 'a t
end

module type ListT1 = T1 with type +'a t := 'a list
```

```
$ ocamlc -c a.ml
$
```
It's checked for consistency with the RHS, too (the above does not work with `-'a t`). But it's not printed faithfully by the pretty printer:

```
$ ocamlc -dsource -c a.ml 
module type T1  = sig type 'a t end
module type ListT1  = T1 with type 'a t :=  'a list
```

AFAICT the source printed this way will always compile, but it's a problem for ppxes and `[@@deriving_inline]`, since ppxlib's `mk_named_sig` will sometimes generate this construction. (It doesn't have to, clearly, so this particular problem could be fixed there as well.)

--- a/Changes
+++ b/Changes
@@ -781,6 +781,10 @@ OCaml 4.14.0
   (Gabriel Scherer and Nathanaëlle Courant,
    review by Florian Angeletti, Ulysse Gérard and Thomas Refis)
 
+- #10790: don't drop variance and injectivity annotations when pretty printing
+  `with` constraints (for example, `with type +!'a t = ...`).
+  (Florian Angeletti, report by Luke Maurer, review by ???)
+
 ### Build system:
 
 - #10835 Disable DT_TEXTREL warnings on x86 32 bit architecture by passing

--- a/Changes
+++ b/Changes
@@ -333,6 +333,11 @@ OCaml 5.0
 
 ### Bug fixes:
 
+- #10790: don't drop variance and injectivity annotations when pretty printing
+  `with` constraints (for example, `with type +!'a t = ...`).
+  (Florian Angeletti, report by Luke Maurer, review by Matthew Ryan and
+   Gabriel Scherer)
+
 - #11167: Fix memory leak from signal stack.
   (Antoni Żewierżejew, review by Gabriel Scherer and Enguerrand Decorne)
 
@@ -780,10 +785,6 @@ OCaml 4.14.0
 - #10742: strong call-by-need reduction for shapes
   (Gabriel Scherer and Nathanaëlle Courant,
    review by Florian Angeletti, Ulysse Gérard and Thomas Refis)
-
-- #10790: don't drop variance and injectivity annotations when pretty printing
-  `with` constraints (for example, `with type +!'a t = ...`).
-  (Florian Angeletti, report by Luke Maurer, review by ???)
 
 ### Build system:
 

--- a/parsing/pprintast.ml
+++ b/parsing/pprintast.ml
@@ -1068,19 +1068,17 @@ and module_type ctxt f x =
 
 and with_constraint ctxt f = function
   | Pwith_type (li, ({ptype_params= ls ;_} as td)) ->
-      let ls = List.map fst ls in
       pp f "type@ %a %a =@ %a"
-        (list (core_type ctxt) ~sep:"," ~first:"(" ~last:")")
-        ls longident_loc li (type_declaration ctxt) td
+        (type_params ctxt) ls
+        longident_loc li (type_declaration ctxt) td
   | Pwith_module (li, li2) ->
       pp f "module %a =@ %a" longident_loc li longident_loc li2;
   | Pwith_modtype (li, mty) ->
       pp f "module type %a =@ %a" longident_loc li (module_type ctxt) mty;
   | Pwith_typesubst (li, ({ptype_params=ls;_} as td)) ->
-      let ls = List.map fst ls in
       pp f "type@ %a %a :=@ %a"
-        (list (core_type ctxt) ~sep:"," ~first:"(" ~last:")")
-        ls longident_loc li
+        (type_params ctxt) ls
+        longident_loc li
         (type_declaration ctxt) td
   | Pwith_modsubst (li, li2) ->
       pp f "module %a :=@ %a" longident_loc li longident_loc li2

--- a/testsuite/tests/parsetree/source.ml
+++ b/testsuite/tests/parsetree/source.ml
@@ -7443,3 +7443,9 @@ let g y =
   f ~(y:int)
 
 let goober a = match a with C (type a b) y -> y
+
+
+(** With constraints *)
+
+module type s = sig type ('a,'b) t end with type (-!'a, !+'b) t = 'b -> 'a list
+module type s = sig type ('a,'b) t end with type (!-'a, +!'b) t := 'b -> 'a list

--- a/testsuite/tests/parsetree/source.ml
+++ b/testsuite/tests/parsetree/source.ml
@@ -7449,3 +7449,4 @@ let goober a = match a with C (type a b) y -> y
 
 module type s = sig type ('a,'b) t end with type (-!'a, !+'b) t = 'b -> 'a list
 module type s = sig type ('a,'b) t end with type (!-'a, +!'b) t := 'b -> 'a list
+module type s = sig type ('a,'b) t end with type ('a,'b) t := 'b -> 'a list

